pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,380 @@
+}
+
+func (c *webhookAuthenticatorCertRequester) syncWithConditions(ctx context.Context, syncCtx factory.SyncContext) error {
+	progressing, err := c.sync(ctx, syncCtx)
+
+	if progressing != nil {
+		var progressingText string
+		progressingStatus := operatorv1.ConditionFalse
+		progressingReason := "CertificateAvailable"
+		if *progressing {
+			progressingText = "waiting for CSR to be signed"
+			progressingStatus = operatorv1.ConditionTrue
+			progressingReason = "WaitingForCertificate"
+		}
+
+		_, _, statusErr := v1helpers.UpdateStatus(c.operatorClient,
+			v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "AuthenticatorCSRProgressing",
+				Status:  progressingStatus,
+				Reason:  progressingReason,
+				Message: progressingText,
+			}))
+
+		if statusErr != nil {
+			if err != nil {
+				klog.Error("failed to update operator status: %v", err)
+				return err
+			}
+			return statusErr
+		}
+	}
+
+	return err
+}
+
+func (c *webhookAuthenticatorCertRequester) sync(ctx context.Context, syncCtx factory.SyncContext) (progressing *bool, err error) {
+	key, cert, keyRotated, err := c.getKeyCertFromSecret(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if keyRotated {
+		if err := c.resubmitCSR(ctx, key); err != nil {
+			return nil, err
+		}
+		return pbool(true), nil
+	}
+
+	if cert != nil {
+		// check that the certificate is not expiring
+		certValidityDuration := cert.NotAfter.Sub(cert.NotBefore)
+		if time.Now().Before(cert.NotBefore.Add(time.Duration(0.8 * float64(certValidityDuration)))) {
+			return pbool(false), nil
+		}
+
+		klog.V(5).Info("the authenticator's certificate is almost expired, will schedule a new CSR creation")
+
+		// delete the cert/key secret and schedule a new sync
+		if err := c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{}); err != nil {
+			return nil, err
+		}
+		return nil, factory.SyntheticRequeueError
+	}
+
+	// either we don't have a cert yet or it's almost expired
+	csrObj, err := c.csrLister.Get(csrName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if csrObj != nil {
+		if len(csrObj.Status.Conditions) == 0 {
+			return pbool(true), nil
+		}
+
+		for _, cond := range csrObj.Status.Conditions {
+			switch cond.Type {
+			case certapiv1.CertificateDenied, certapiv1.CertificateFailed:
+				if cond.Status == corev1.ConditionTrue {
+					syncCtx.Recorder().Eventf("CSRFailure", "previous CSR got denied or it otherwise failed: {%s - %s: %s}", cond.Type, cond.Reason, cond.Message)
+					klog.Warning("the previous CSR got denied or it otherwise failed: %v", cond)
+					if err := c.resubmitCSR(ctx, key); err != nil {
+						return nil, err
+					}
+					return pbool(true), nil
+				}
+			}
+		}
+
+		for _, cond := range csrObj.Status.Conditions {
+			if cond.Type == certapiv1.CertificateApproved && cond.Status == corev1.ConditionTrue {
+				csrCert := csrObj.Status.Certificate
+				if len(csrCert) == 0 {
+					return pbool(true), nil
```

Comment:
would be nice if we could store progress in the status, like `csr got approved, waiting for cert`

Reply:
I did not want to overcomplicate the condition handling, because I expect we would see a very very fast (about 1s each state) succession of:
1. I submitted the CSR
2. The CSR is pending
3. The CSR got approved, waiting for cert
4. AsExpected